From 9d7833a6a7be78e02eb8132833829a2baa67d8ab Mon Sep 17 00:00:00 2001
From: Zeng Heng <zengheng4@huawei.com>
Date: Mon, 24 Feb 2025 16:51:01 +0800
Subject: [PATCH 694/960] downstream: arm64/mpam: Try reading again if monitor
 instance returns not ready

If the MSC monitor instance is the first time to configure, this single
monitor need to wait 'not ready' time before getting the valid value.

Components made up of multiple MSC may need values from each MSC, so when
getting the whole component monitor values, need to wait every single
monitor instance 'not ready' time instead of once per component.

Fixes: bb66b4d115e5 ("arm_mpam: Add mpam_msmon_read() to read monitor value")
Signed-off-by: Zeng Heng <zengheng4@huawei.com>
Signed-off-by: Liu Shixin <liushixin2@huawei.com>
---
 drivers/platform/mpam/mpam_devices.c | 37 +++++++++++++---------------
 1 file changed, 17 insertions(+), 20 deletions(-)

diff --git a/drivers/platform/mpam/mpam_devices.c b/drivers/platform/mpam/mpam_devices.c
index e10be1accd75..ec981994ec8e 100644
--- a/drivers/platform/mpam/mpam_devices.c
+++ b/drivers/platform/mpam/mpam_devices.c
@@ -1031,11 +1031,14 @@ static void __ris_msmon_read(void *arg)
 	}
 
 	*(m->val) += now;
+	m->err = 0;
 }
 
 static int _msmon_read(struct mpam_component *comp, struct mon_read *arg)
 {
 	int err, idx;
+	bool read_again;
+	u64 wait_jiffies;
 	struct mpam_msc *msc;
 	struct mpam_msc_ris *ris;
 
@@ -1044,10 +1047,23 @@ static int _msmon_read(struct mpam_component *comp, struct mon_read *arg)
 		arg->ris = ris;
 
 		msc = ris->msc;
+		read_again = false;
+again:
 		mutex_lock(&msc->lock);
 		err = smp_call_function_any(&msc->accessibility,
 					    __ris_msmon_read, arg, true);
 		mutex_unlock(&msc->lock);
+
+		if (arg->err == -EBUSY && !read_again) {
+			read_again = true;
+
+			wait_jiffies = usecs_to_jiffies(comp->class->nrdy_usec);
+			while (wait_jiffies)
+				wait_jiffies = schedule_timeout_uninterruptible(wait_jiffies);
+
+			goto again;
+		}
+
 		if (!err && arg->err)
 			err = arg->err;
 		if (err)
@@ -1061,9 +1077,7 @@ static int _msmon_read(struct mpam_component *comp, struct mon_read *arg)
 int mpam_msmon_read(struct mpam_component *comp, struct mon_cfg *ctx,
 		    enum mpam_device_features type, u64 *val)
 {
-	int err;
 	struct mon_read arg;
-	u64 wait_jiffies = 0;
 	struct mpam_props *cprops = &comp->class->props;
 
 	might_sleep();
@@ -1080,24 +1094,7 @@ int mpam_msmon_read(struct mpam_component *comp, struct mon_cfg *ctx,
 	arg.val = val;
 	*val = 0;
 
-	err = _msmon_read(comp, &arg);
-	if (err == -EBUSY)
-		wait_jiffies = usecs_to_jiffies(comp->class->nrdy_usec);
-
-	while (wait_jiffies)
-		wait_jiffies = schedule_timeout_uninterruptible(wait_jiffies);
-
-	if (err == -EBUSY) {
-		memset(&arg, 0, sizeof(arg));
-		arg.ctx = ctx;
-		arg.type = type;
-		arg.val = val;
-		*val = 0;
-
-		err = _msmon_read(comp, &arg);
-	}
-
-	return err;
+	return _msmon_read(comp, &arg);
 }
 
 void mpam_msmon_reset_all_mbwu(struct mpam_component *comp)
-- 
2.43.0

